### ISSUE TITLE
AutoRest does not load the correct version of extension '@autorest/csharp'

### ISSUE THREAD

**Issue by pshao25** [Azure/azure-sdk-for-net#35230]
Repro steps:
1. At any time of this repo, clean all the cache, for example at that time the `Microsoft.Azure.AutoRest.CSharp` version is `3.0.0-beta.20230328.2`.
2. Run `dotnet msbuild /t:GenerateCode` under an arbitrary project. The version autorest is loading should be `3.0.0-beta.20230328.2`.
3. Change the `Microsoft.Azure.AutoRest.CSharp` version to `3.0.0-beta.20230329.1`.
4. Run `dotnet msbuild /t:GenerateCode` again, you would see the `Microsoft.Azure.AutoRest.CSharp` version is still `3.0.0-beta.20230328.2` instead of `3.0.0-beta.20230329.1`.
>  info    |    Loading local AutoRest extension '@autorest/csharp' (C:/Users/pashao.FAREAST/.nuget/packages/microsoft.azure.autorest.csharp/3.0.0-beta.20230328.2/buildMultiTargeting/../tools/net6.0/any/)

Root cause: 
You could go to `{RopoRoot}\artifacts\obj\Azure.Analytics.Synapse.Monitoring` for example, there is a cache file `project.nuget.cache`. Its content is 
![image](https://user-images.githubusercontent.com/97225342/228775008-393286a6-9ddc-435b-9fd2-57d6aec75c5c.png)

Suggestion:
I suggest we add a `dotnet restore` to every `autorest.md`, otherwise user might generate incorrect code from out-of-date generator.
![image](https://user-images.githubusercontent.com/97225342/228775636-6a846fee-a698-49f9-9c63-7b77de23db60.png)

**Comment by jsquire** [comment]
//cc: @m-nash, @heaths, pallavit

**Comment by m-nash** [comment]
I think this is a common thing for many library dependency cache situations.  If we wanted to fix this I would lean more towards having the cdoe generation target always run restore for the user so they don't need to run multiple commands.

**Comment by pallavit** [comment]
Is this issue more of a concern now than before due to all the activity happening in the autorest repo? As Michael said I would assume this is package caching concern which is usually diminished by backward compatibility. So, if you end up using an older package you generate older code but not necessarily incorrect code.

**Comment by heaths** [comment]
I agree with Michael. Solve this as low down the stack as possible by depending on the `Restore` target for `GenerateCode`. You might need to find some way to force it, though. Maybe. Given a `<PackageReference>` changed that may not be necessary.

If you fix it in some higher-level script, if and when someone calls the `GenerateCode` target themselves or another script is written for some other purpose, the problem will be apparent again and may be hard to diagnose for people who aren't as familiar with the system end-to-end.

**Comment by pshao25** [comment]
> Is this issue more of a concern now than before due to all the activity happening in the autorest repo? As Michael said I would assume this is package caching concern which is usually diminished by backward compatibility. So, if you end up using an older package you generate older code but not necessarily incorrect code.

I would rephrase as "user might generate unexpected code from out-of-date generator."

**Comment by jsquire** [comment]
Per discussion, if this is to be addressed, it should be done lower in the stack, such as part of the generation process.  As AutoRest is not really under active development and there's been new follow-up or feedback requesting this, I'm going to close this out.
